x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/UnigramTokenizer.java

```diff
@@ -0,0 +1,489 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.nlp.tokenizers;
+
+import org.apache.lucene.analysis.CharArraySet;
+import org.apache.lucene.analysis.CharacterUtils;
+import org.apache.lucene.analysis.Tokenizer;
+import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
+import org.apache.lucene.analysis.tokenattributes.OffsetAttribute;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.UnicodeUtil;
+import org.elasticsearch.common.util.Maps;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+import static org.elasticsearch.xpack.ml.inference.nlp.tokenizers.TokenizerUtils.numUtf8Bytes;
+import static org.elasticsearch.xpack.ml.inference.nlp.tokenizers.TokenizerUtils.splitOutNeverSplit;
+
+/**
+ * Sentence-piece unigram tokenizer.
+ *
+ * Does whitespace tokenization with unigram tokenization on the resulting tokens.
+ *
+ * This cannot be a token-filter as it needs access to the offset correction logic provided by the upstream CharFilter.
+ *
+ * You may notice that the offsets are always matching the individual tokens position back to the original string. This is because
+ * there aren't "sub-word" tokens, per-se. So, we don't have tokens that share the same offsets as in WordPiece.
+ */
+public final class UnigramTokenizer extends Tokenizer {
+    private static final double K_UNK_PENALTY = 10.0;
+    static final String PREFIX = "▁";
+
+    private final CharTermAttribute termAtt = addAttribute(CharTermAttribute.class);
+    private final OffsetAttribute offsetAtt = addAttribute(OffsetAttribute.class);
+
+    static UnigramTokenizer build(List<String> neverSplit, List<String> dictionary, List<Double> scores, String unknownToken) {
+        if (dictionary.isEmpty()) {
+            throw new IllegalArgumentException("vocab empty");
+        }
+        if (unknownToken == null) {
+            throw new IllegalArgumentException("unknown token ID");
+        }
+        CharArraySet neverSplitSet = new CharArraySet(neverSplit, false);
+        CharTrie neverSplitTree = CharTrie.build(neverSplit);
+        if (dictionary.size() != scores.size()) {
+            throw new IllegalArgumentException("Yo! These should be equal!!!");
```

Comment:
Needs a proper error message, and also include both sizes to help debugging if this ever happens.

Reply:
Ah, yeah! My bad :D